src/main/kotlin/cacophony/grammars/Grammar.kt

```diff
@@ -1,13 +1,27 @@
 package cacophony.grammars
 
+import cacophony.automata.joinAutomata
+import cacophony.automata.minimalization.buildDFAFromRegex
 import cacophony.utils.AlgebraicRegex
 
 data class Production<SymbolType>(
     val lhs: SymbolType,
     val rhs: AlgebraicRegex<SymbolType>,
 )
 
+infix fun <SymbolType> SymbolType.produces(regex: AlgebraicRegex<SymbolType>) = Production(this, regex)
+
 data class Grammar<SymbolType>(
     val start: SymbolType,
     val productions: Collection<Production<SymbolType>>,
-)
+) {
+    fun buildAutomata() = productions.groupBy { it.lhs }.mapValues { buildAutomaton(it.value) }
+
+    private fun buildAutomaton(productions: List<Production<SymbolType>>) =
+        joinAutomata(productions.map { buildDFAFromRegex(it.rhs) }.zip(productions))
+}
+
+fun <SymbolType> grammarOf(
+    start: SymbolType,
+    vararg productions: Production<SymbolType>,
+) = Grammar(start, productions.toList())
```

Comment:
i'd stick to the convention of companion objects and `Grammar::fromSomething()` factory methods

this one could be for example `::fromProductions()`

i feel like `somethingOf()` is tied to containers

Reply:
I removed this completely since we can just use the constructor `Grammar(start, listOf(...))`